### PR TITLE
[Matrix] change to new C++ PVR interface way

### PR DESCRIPTION
**Pull request by AlwinEsch**
Changes related to xbmc/xbmc#16485.

This make the addon to the new C++ addon interface.

With missing backend account not runtime tested!

### CODE REVIEW

File: src/Data.cpp [hunk 1]
```diff
@@ -100,21 +92,14 @@ Data::Data(Configuration cfg)
...
-  , m_useH265{cfg.useH265}
-  , m_useAdaptive{cfg.useAdaptive}
-  , m_showLockedChannels{cfg.showLockedChannels}
-  , m_showLockedOnlyPin{cfg.showLockedOnlyPin}
-  , m_currentStreamIsLive{false}
-  , m_manager{std::move(cfg.userName), std::move(cfg.password), std::move(cfg.deviceId), std::move(cfg.productId)}
+  , m_manager{std::move(kodi::GetSettingString("userName")),
+              std::move(kodi::GetSettingString("password")),
+              std::move(kodi::GetSettingString("deviceId")),
+              std::move(kodi::GetSettingString("productId"))}
```
Comment (AlwinEsch) [inline src/Data.cpp]: About here maybe not the best? 

Addon load test was OK with them, but as the external class creation is no more present, was the **"struct Configuration"** not usable on his constructor and added by direct read.

File: src/Data.cpp [hunk 2]
```diff
@@ -355,7 +345,78 @@ Data::~Data(void)
...
+
+  m_streamQuality = kodi::GetSettingEnum<ApiManager::StreamQuality_t>("streamQuality", ApiManager::SQ_DEFAULT);
+  m_fullChannelEpgRefresh = kodi::GetSettingInt("fullChannelEpgRefresh", 24) * 3600; // make it seconds
+  m_loadingsRefresh = kodi::GetSettingInt("loadingsRefresh", 60);
+  m_keepAliveDelay = kodi::GetSettingInt("keepAliveDelay", 20);
+  m_epgCheckDelay = kodi::GetSettingInt("epgCheckDelay", 1) * 60; // make it seconds
+  m_useH265 = kodi::GetSettingBoolean("useH265", false);
+  m_useAdaptive = kodi::GetSettingBoolean("useAdaptive", false);
+  m_showLockedChannels = kodi::GetSettingBoolean("showLockedChannels", true);
+  m_showLockedOnlyPin = kodi::GetSettingBoolean("showLockedOnlyPin", true);
```
Comment (AlwinEsch) [inline src/Data.cpp:368]: Here the other setting values are taken and stored in class.

File: src/Data.cpp [hunk 3]
```diff
@@ -288,9 +272,15 @@ void Data::LoginLoop()
     if (0 >= login_delay)
     {
       if (m_manager.login())
+      {
+        ConnectionStateChange("Connected", PVR_CONNECTION_STATE_CONNECTED, "");
         should_try = false;
+      }
       else
+      {
+        ConnectionStateChange("Disconnected", PVR_CONNECTION_STATE_DISCONNECTED, "");
```
Comment (AlwinEsch) [inline src/Data.cpp:281]: This "ConnectionStateChange" is common to use on addon, as the call **`ADDON_GetStatus()`** no more used since longer time.

Here it handle only the minimal about login, maybe it can also added on other places to inform Kodi about changes.
See http://alwinesch.github.io/group__cpp__kodi__addon__pvr___defs___general___p_v_r___c_o_n_n_e_c_t_i_o_n___s_t_a_t_e.html#gaa434a9da21b265c38eade0b452d8d58b

Note: The text e.g. "Disconnected" is not really used on Kodi, only added to have something on this value :smile: 
It takes normally the last two, if the 3. is empty it generate a internal message about.

Here what comes about this:
Bildschirmfoto vom 2020-06-04 18-49-06

**Comment by palinek** [comment]
Great job @AlwinEsch. Still compiling the xbmc/xbmc#16485 and reviewing code.. I'll let you know ASAP...

File: src/Data.cpp [hunk 4]
```diff
@@ -788,18 +840,37 @@ PVR_ERROR Data::GetChannels(ADDON_HANDLE handle, bool bRadio)
...
+    return ret;
+
+  StreamProperties(properties, streamUrl, streamType, true);
+  return PVR_ERROR_NO_ERROR;
+}
+
+PVR_ERROR Data::GetSignalStatus(int channelUid, kodi::addon::PVRSignalStatus& signalStatus)
+{
+  signalStatus.SetAdapterName("sledovanitv.cz");
+  signalStatus.SetAdapterStatus("OK");
```
Comment (palinek) [inline src/Data.cpp:857]: What are possible values here?

File: src/Data.cpp [hunk 5]
```diff
@@ -1240,27 +1357,18 @@ bool Data::LoggedIn() const
   return m_manager.loggedIn();
 }
 
-properties_t Data::StreamProperties(const std::string & url, const std::string & streamType, bool isLive)
+void Data::StreamProperties(std::vector<kodi::addon::PVRStreamProperty>& properties, const std::string & url, const std::string & streamType, bool isLive)
```
Comment (palinek) [inline src/Data.cpp]: Why did you move the return type to be parameter?

File: src/Data.cpp [hunk 6]
```diff
@@ -100,21 +92,14 @@ Data::Data(Configuration cfg)
...
-  , m_loadingsRefresh{cfg.loadingsRefresh}
-  , m_keepAliveDelay{cfg.keepAliveDelay}
-  , m_epgCheckDelay{cfg.epgCheckDelay}
-  , m_useH265{cfg.useH265}
-  , m_useAdaptive{cfg.useAdaptive}
-  , m_showLockedChannels{cfg.showLockedChannels}
-  , m_showLockedOnlyPin{cfg.showLockedOnlyPin}
-  , m_currentStreamIsLive{false}
-  , m_manager{std::move(cfg.userName), std::move(cfg.password), std::move(cfg.deviceId), std::move(cfg.productId)}
+  , m_manager{std::move(kodi::GetSettingString("userName")),
```
Comment (palinek) [inline src/Data.cpp]: I'm not certain, but doesn't `std::move` prohibit copy elision here?

**Comment by palinek** [comment]
Few tests with first run...  everything seems to be working as expected.

File: src/Data.cpp (hunk 6 above)
Comment (AlwinEsch) [reply src/Data.cpp]: Also not sure, thing to remove it, normally return settings a created "std::string" like with "cfg" before, but with "const std::string &" as value on class makes itnot direct sense.

File: src/Data.cpp (hunk 5 above)
Comment (AlwinEsch) [reply src/Data.cpp]: Was to remove some not needed malloc's and free's in C++ background, as by return it recreate all data to store on his parent call and delete the ones from return.

File: src/Data.cpp (hunk 4 above)
Comment (AlwinEsch) [reply src/Data.cpp:857]: Possible values are visible here http://alwinesch.github.io/group__cpp__kodi__addon__pvr___defs___channel___p_v_r_signal_status.html

File: src/Data.cpp (hunk 5 above)
Comment (palinek) [reply src/Data.cpp]: > as by return it recreate all data to store on his parent call and delete the ones from return.

This is definitely not the case up from C++11 (move constructors etc..)

**Comment by AlwinEsch** [comment]
Have it updated, the `std::move` from constructor is away and the `std::vector` on properties as return.

**Comment by palinek** [comment]
Closes #64

**Comment by palinek** [comment]
@AlwinEsch A big **Thank you**, I really appreciate, that you've done all the "boring" work here

**Comment by AlwinEsch** [comment]
No problem, it's fun and good hobby work, even if that was a lot with the many addons :smile:
